models/user/user_test.go

```diff
@@ -544,3 +544,24 @@ func Test_ValidateUser(t *testing.T) {
 		assert.EqualValues(t, expected, err == nil, fmt.Sprintf("case: %+v", kase))
 	}
 }
+
+func Test_NormalizeUserFromEmail(t *testing.T) {
+	testCases := [][]interface{}{
+		// input name, expected normalized name, is normalized name valid
+		{"test", "test", true},
+		{"Sinéad.O'Connor", "Sin-ad.OConnor", true}, // We should consider allowing custom replacement characters (eg. é -> e)
```

Comment:
I guess it's necessary to do so .... maybe the Go library also has ability to do so.

Otherwise `Sin-ad` doesn't seem good for a username.

Reply:
Another reason why it's necessary: maybe nobody likes making "breaking" changes again and again.

Think about a case: if this PR doesn't have a complete solution, in the future, if someone decides to "consider allowing replacement characters", what they could do? They needs to do introduce "option_v2", "option_v3" .... it would be a mess.

Reply:
[There was some discussion in discord](https://discord.com/channels/322538954119184384/322538954119184384/1184568790948774029) about this too.

I'm not sure that we want to take on normalizing every possible user input the way that a user would want/expect. The most accessible method would be to extend this solution in the future with either another config setting, like `NORMALIZE_CHARS`, or add it directly to the auth source form. 

However... if we want to take it on, there is a pretty good lookup table here we could transpose from javascript: https://web.archive.org/web/20120918093154/http://lehelk.com/2011/05/06/script-to-remove-diacritics/

I'm happy to go either way, but if we do the first method, we can follow up with a future, non-breaking PR.

I guess the problem with the first option is that admins could easily break user accounts if they change the character replacements and then restart the server.

Reply:
The unicode standard has defined it.

https://go.dev/blog/normalization

Reply:
That article is specifically for characters that can be decomposed with diacritics. So, it should work for characters with accents (though the code they provide is out of date and needs to be modified), but is not a general-solution nor a complete "normalization" imo.

I think there is a good reason other software companies don't do this ([eg](https://docs.sourcegraph.com/@v3.39.1/admin/auth#username-normalization).): we are now making some assumptions about how users' names should be transformed, and it's only one step in that direction, which could potentially be endless. For example, `Ægidius` will still become `-gidius`, which will fail user creation. 

After putting some thought into it I agree that we should the change you suggest. However, if we choose to go this direction I suggest we do NOT replace unknown characters with `-`. The reasons being:

- As you say, that could result in breaking changes in the future if we decide to replace additional characters. Instead, we selectively add characters into the replacement set safely.
- This solves all of the open issues, while providing significant, incremental improvement